### PR TITLE
WT-7872 Replace tiered storage use of WT_FH handles with WT_BLOCK handles

### PR DESCRIPTION
**Pull request by keithbostic**
Replace tiered storage use of WT_FH handles with WT_BLOCK handles.

@fedorova, this impacts the structure of your work in caching.

### CODE REVIEW

File: src/block/block_addr.c [hunk 1]
```diff
@@ -113,7 +113,7 @@ __wt_block_addr_pack(WT_BLOCK *block, uint8_t **pp, uint32_t objectid, wt_off_t
      * starting with a single object with no object IDs, where all future objects in the stack know
      * a missing object ID is a reference to the base object.
      */
-    if (i != 0 && block->has_objects) {
+    if (i != WT_BLOCK_OBJECTID_NONE) {
```
Comment (sueloverso) [inline src/block/block_addr.c:116]: It seems like this could be useful anywhere object ids are used. I suggest naming it `WT_TIERED_OBJECTID_NONE` and putting it in `tiered.h` instead. There is a line in `tiered_handle.c` that checks `tiered->current_id != 0` too.

File: src/block/block_open.c [hunk 2]
```diff
@@ -82,33 +82,31 @@ __wt_block_manager_create(WT_SESSION_IMPL *session, const char *filename, uint32
 }
 
 /*
- * __block_destroy --
+ * __wt_block_close --
  *     Destroy a block handle.
```
Comment (sueloverso) [inline src/block/block_open.c:86]: Should this now say "Close a block handle"?

File: src/block/block_open.c [hunk 3]
```diff
@@ -210,17 +209,19 @@ __wt_block_open(WT_SESSION_IMPL *session, const char *filename, WT_BLOCK_FILE_OP
     else if (WT_STRING_MATCH("sequential", cval.str, cval.len))
         LF_SET(WT_FS_OPEN_ACCESS_SEQ);
 
+    if (fixed)
+        LF_SET(WT_FS_OPEN_FIXED);
```
Comment (sueloverso) [inline src/block/block_open.c:213]: This seems unrelated? Is this part of some other ticket? (Also should be alphabetized.)

File: src/block_cache/block_tier.c [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    cfg[0] = object_val;
+    cfg[1] = NULL;
+
+    /* Check if the object exists. */
+    exist = true;
+    if (!local_only)
+        WT_ERR(__wt_fs_exist(session, object_name, &exist));
+    if (exist)
+        WT_ERR(
+          __wt_block_open(session, object_name, objectid, cfg, false, readonly, false, 0, &block));
```
Comment (sueloverso) [inline src/block_cache/block_tier.c:73]: It seems possible that between the existence check and opening the local file on an older object that the object could have been locally dropped (i.e. add a delay after line 70). I think if `wt_block_open` returns an error in the `!local_only` case (i.e. the local file disappeared) then instead of an error the code should try opening in the bucket.

**Review by sueloverso — APPROVED** [review]
LGTM with a few questions and comments.

File: src/block/block_addr.c (hunk 1 above)
Comment (keithbostic) [reply src/block/block_addr.c]: Fixed in c03722407.

File: src/block/block_open.c (hunk 2 above)
Comment (keithbostic) [reply src/block/block_open.c]: Fixed in 8f571b2e9.

File: src/block/block_open.c (hunk 3 above)
Comment (keithbostic) [reply src/block/block_open.c:213]: This wasn't an unrelated change, tiered storage opens fixed filename paths. (I have no idea if that's right or not, but that's what it does from before, see https://github.com/wiredtiger/wiredtiger/blob/develop/src/tiered/tiered_opener.c#L73).

I'm not sure what you mean by "should be alphabetized"? Can you explain further, please?

File: src/block_cache/block_tier.c (hunk 4 above)
Comment (keithbostic) [reply src/block_cache/block_tier.c:73]: This is what @ddanderson was talking about on Slack; there has to be some locking or other verification during open in order for the open and the remove not to race. As far as I know, I've not made this better or worse, it's as broken as it always ways.

Comment (sueloverso) [reply src/block_cache/block_tier.c:73]: Well, the race didn't really exist before because we were only calling `wt_open` and not checking existence beforehand, so there was no window. In the old code, if we couldn't open the local (but older) object for any reason we tried opening the bucket object.

File: src/block/block_open.c (hunk 3 above)
Comment (sueloverso) [reply src/block/block_open.c:213]: By "alphabetized" I mean that these two lines should go below the ones that set WT_FS_OPEN_DIRECTIO a few lines down.

File: src/block_cache/block_tier.c (hunk 4 above)
Comment (keithbostic) [reply src/block_cache/block_tier.c:73]: The previous race was an open followed by a remove which will fail on Windows.

File: src/block/block_open.c (hunk 3 above)
Comment (keithbostic) [reply src/block/block_open.c:213]: That would mean moving the lines into the middle of 3 clauses, all dealing with the `readonly` flag, that's why I did it the way I did.

(The fact we're passing 3 booleans into this function is flawed; we should rework this flag handling entirely.)

File: src/block_cache/block_mgr.c [hunk 5]
```diff
@@ -494,22 +557,55 @@ __bm_stat(WT_BM *bm, WT_SESSION_IMPL *session, WT_DSRC_STATS *stats)
...
+     *  - close this object (about to move it, don't allow reopens yet)
+     *  - allow opens on this object again
+     */
+    WT_UNUSED(flags);
+    WT_RET(__wt_blkcache_get_handle(session, NULL, objectid, &block));
+
+    /*
+     * KEITH XXX: We need to distinguish between tiered switch and loading a checkpoint. This is
+     * also discarding the extent list which isn't correct, because we can't know when to discard
+     * previous files if we don't have the extent list. This fixes the problem where we randomly
```
Comment (keitharnoldsmith) [inline src/block_cache/block_mgr.c:587]: Is this something we are comfortable with as future work if we accept this change?  Or is this an unknown that we need to figure out how to address before accepting this change?

File: src/block_cache/block_tier.c [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    object_uri = object_val = NULL;
+
+    WT_ASSERT(session, objectid <= tiered->current_id);
+    WT_ASSERT(session, uri == NULL || WT_PREFIX_MATCH(uri, "tiered:"));
+    WT_ASSERT(session, (uri == NULL && objectid != 0) || (uri != NULL && objectid == 0));
+
+    /*
+     * First look for the local file. This will be the fastest access and we retain recent objects
+     * in the local database for awhile. If we're passed a name to open, then by definition it's a
+     * local file.
```
Comment (keitharnoldsmith) [inline src/block_cache/block_tier.c:40]: The recent discussion of coordination (or lack of it) around removing files we've retained in this manner makes me wonder if this is something we want as part of the long term design.  We know that eventually we are going to want a general purpose way to cache data from object storage on a local file system.  Should we take some time to decide how that is going to work and whether/how it pulls data into that cache after we flush a file to object storage before we worry about addressing the file removal race?

Just dropping this note here since this comment triggered the thought and all of the interested folks are reviewing the PR.  This is clearly outside the scope of this change.

File: src/block_cache/block_mgr.c [hunk 7]
```diff
@@ -182,6 +182,69 @@ __bm_checkpoint_unload(WT_BM *bm, WT_SESSION_IMPL *session)
     return (ret);
 }
 
+/*
+ * __bm_close_block_remove --
+ *     Remove a single block.
```
Comment (keitharnoldsmith) [inline src/block_cache/block_mgr.c:187]: Is there an alternate name for a `WT_BLOCK`? Might people be confused because we also talk about reading and writing individual blocks?  

This is probably baked into WT code and terminology. So feel free to ignore if I'm howling at the moon here.

**Review by keitharnoldsmith — COMMENTED** [review]
In general this is looking good.  I'm submitting now to share some initial comments before I go into a string of evening meetings.

File: src/block_cache/block_mgr.c [hunk 8]
```diff
@@ -182,6 +182,69 @@ __bm_checkpoint_unload(WT_BM *bm, WT_SESSION_IMPL *session)
...
+static int
+__bm_close_block_remove(WT_SESSION_IMPL *session, WT_BLOCK *block)
+{
+    u_int i;
+
+    __wt_verbose(session, WT_VERB_BLOCK, "close: %s", block->name);
+
+    /* Discard any references we're holding. */
+    for (i = 0; i < block->related_next; ++i) {
+        --block->related[i]->ref;
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/block_cache/block_mgr.c:198]: Yes, they're block local, which is a single-threaded structure.

File: src/block_cache/block_mgr.c [hunk 9]
```diff
@@ -182,6 +182,69 @@ __bm_checkpoint_unload(WT_BM *bm, WT_SESSION_IMPL *session)
...
+static int
+__bm_close_block(WT_SESSION_IMPL *session, WT_BLOCK *block)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    bool found;
+
+    conn = S2C(session);
+
+    /* You can't close  files during a checkpoint. */
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/block_cache/block_mgr.c]: Fixed in 525813135.

File: src/block_cache/block_mgr.c (hunk 5 above)
Comment (keithbostic) [reply src/block_cache/block_mgr.c:587]: @ddanderson should comment, but I believe this comment simply notes the current state of the tiered storage project: tiered storage can't create "valid" database objects yet, and switching between writeable files is part of that.

File: src/block_cache/block_mgr.c (hunk 7 above)
Comment (ddanderson) [reply src/block_cache/block_mgr.c:187]: +1 this was confusing for me getting started with this code.

File: src/block_cache/block_tier.c (hunk 6 above)
Comment (keithbostic) [reply src/block_cache/block_tier.c:40]: There are a variety of options here -- for example, you could cache data from object storage, in DRAM, NVRAM, the OS buffer cache, or local files, and it can be at either the file-level or individual block level. It's a big discussion.

File: src/block_cache/block_mgr.c (hunk 7 above)
Comment (keithbostic) [reply src/block_cache/block_mgr.c]: Fixed in ccc845096.

At the moment, `WT_BM` is a block manager a.k.a "file", and `WT_BLOCK` is a block interface handle, a.k.a. "file", so this needs cleanup. They should probably merge because there's no reason to believe different objects will have the same methods for access.

Comment (ddanderson) [reply src/block_cache/block_mgr.c]: Good to pursue in a different ticket, I think.

File: src/block/block_open.c [hunk 10]
```diff
@@ -82,33 +82,31 @@ __wt_block_manager_create(WT_SESSION_IMPL *session, const char *filename, uint32
...
     WT_CONNECTION_IMPL *conn;
     WT_DECL_RET;
-    uint64_t bucket;
-    u_int i;
+    uint64_t bucket, hash;
 
     conn = S2C(session);
-    bucket = block->name_hash & (conn->hash_size - 1);
+
+    __wt_verbose(session, WT_VERB_BLOCK, "remove: %s", block->name == NULL ? "" : block->name);
```
Comment (fedorova) [inline src/block/block_open.c:97]: The wording is a bit inconsistent. The function name got changed to "block_close", but the verbose message says that we are removing the handle.

Comment (keithbostic) [reply src/block/block_open.c:97]: Fixed in 6ced72286.

File: src/block_cache/block_mgr.c (hunk 7 above)
Comment (fedorova) [reply src/block_cache/block_mgr.c]: +1 Same here. I've been confused about this since I saw this terminology.

File: src/block_cache/block_tier.c (hunk 6 above)
Comment (fedorova) [reply src/block_cache/block_tier.c:40]: I'd love to be in the loop with respect to caching discussions.

**Review by fedorova — APPROVED** [review]

File: src/block_cache/block_mgr.c (hunk 7 above)
Comment (keithbostic) [reply src/block_cache/block_mgr.c]: FTR, I'm likely to simply merge the `WT_BM` structure into the `WT_BLOCK` structure. I see both the block cache and block layers as "serving blocks", so I think `WT_BLOCK` makes sense.

File: src/block_cache/block_tier.c (hunk 6 above)
Comment (keitharnoldsmith) [reply src/block_cache/block_tier.c:40]: @keithbostic : in this context I'm specifically thinking about caching in local storage.  But I agree we should be conscious of the other options because inevitably we'll want some (all?) of them in the future.

@fedorova I'd love to have your input when we get to caching.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM!

File: src/block_cache/block_mgr.c [hunk 11]
```diff
@@ -182,6 +182,69 @@ __bm_checkpoint_unload(WT_BM *bm, WT_SESSION_IMPL *session)
...
+    WT_DECL_RET;
+    bool found;
+
+    __wt_verbose(session, WT_VERB_BLKCACHE, "close: %s", block->name);
+
+    conn = S2C(session);
+
+    /* You can't close files during a checkpoint. */
+    WT_ASSERT(
+      session, block->ckpt_state == WT_CKPT_NONE || block->ckpt_state == WT_CKPT_PANIC_ON_FAILURE);
```
Comment (ddanderson) [inline src/block_cache/block_mgr.c:221]: Do we need to take a checkpoint lock here to really prevent this?

**Review by ddanderson — APPROVED** [review]
LGTM - this looks like a great improvement.

File: src/block_cache/block_mgr.c (hunk 11 above)
Comment (keithbostic) [reply src/block_cache/block_mgr.c:221]: I don't think so -- I put this in as a debugging hack because I got the reference counting wrong, and figured I'd leave it to guard against future missteps.
